src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs

```diff
@@ -28,6 +28,7 @@ internal abstract class DictionaryDefaultConverter<TDictionary, TKey, TValue>
             if (state.Current.CollectionEnumerator == null)
             {
                 enumerator = value.GetEnumerator();
+                state.Current.CollectionEnumerator = enumerator;
```

Comment:
Because this is now setting the enumerator early on, you should probably remove the assignments happening later in the same method bodies.

Reply:
Might be worth checking if this early assignment could break invariants elsewhere in the serializer (look for possible locations where `CollectionEnumerator is null` is being checked)

Reply:
I have cleaned up these duplicate assignments.
Unfortunately, the enumerator for the `DictionaryOfTKeyTValueConverter` class  has the `Dictionary<TKey, TValue>.Enumerator` type which a _value type_. I have changed it to `IEnumerator<KeyValuePair<TKey, TValue>>` to work with a _reference type_ but my first performance tests reveal that it introduces some noticeable performance regressions :/
Do you have any idea how to tackle this issue?

---

I have not found any broken invariant for `CollectionEnumerator`.

Reply:
In most built-in collections and compiler generated iterators the dispose method is a no-op anyway. Perhaps it makes sense to precompute whether that is the case and only store the `CollectionEnumerator` early if found to be a custom enumerator or an enumerator known to have dispose logic.

Reply:
Ok, I have removed the early assignment in the `DictionaryOfTKeyTValueConverter` class as it is the only one in such a situation.

Reply:
I don't think that would work here, the `TCollection` could be any IDictionary type which may or may not have a disposable enumerator. It might make sense to just check against a list of known types, although that would probably require new API for the case of the source generator.

In any case, a fix seems to be creating more problems than what it is solving :-(

Reply:
Here `TCollection` inherits from `Dictionary<TKey, TValue>` where the `GetEnumerator()` method is _not_ virtual, so I would not expect any issue.
Do I miss something?

Reply:
TDictionary inherits from IEnumerable, not Dictionary. The enumerator may or may not require disposal.

Reply:
> where we can safely ignore the disposal of the enumerator as it is implemented as a [no-op](https://github.com/dotnet/runtime/blob/c1a4c9c48f215547679b49f8e2fce5077a63f945/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L1418) and cannot be overridden.

That's not quite true, the converter applies to types inheriting from Dictionary which may or may not be implementing disposal semantics in the enumerator.